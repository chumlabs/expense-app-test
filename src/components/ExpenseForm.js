// IMPORTS
// - libraries
import React, { Component } from 'react';
import moment from 'moment';
import 'react-dates/initialize'; // needed for react-dates
import { SingleDatePicker } from 'react-dates';
import { Redirect } from 'react-router-dom';

export default class ExpenseForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toFixed(2).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = date => {
    // note: date is a moment object
    // if (date) {
    //   this.setState(() => ({ createdAt: date }));
    // }
    this.setState(() => ({ createdAt: date }));
    // console.log(`change to date: `, date);
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
    // optional validation
    // if (!this.state.createdAt) {
    //   this.setState(() => ({ error: 'Please enter a valid date' }));
    // } else {
    //   this.setState(() => ({ error: '' }));
    // }
  };

  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onSubmit = e => {
    e.preventDefault();

    if (!this.state.description) {
      this.setState(() => ({ error: 'Please enter a description' }));
    } else if (!this.state.amount) {
      this.setState(() => ({ error: 'Please enter an amount' }));
    } else if (!this.state.createdAt) {
      this.setState(() => ({ error: 'Please enter a valid date' }));
    } else {
      this.setState(() => ({ error: '' }));
      // console.log('submitted!');
      // use onSubmit prop
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <form onSubmit={this.onSubmit} className="form-group">
        <div className="form-group__error">
          {this.state.error ? <p className="error-text">{this.state.error}</p> : null}
        </div>
        <input
          className="form-group__text-input"
          type="text"
          placeholder="Description (required)"
          value={this.state.description}
          onChange={this.onDescriptionChange}
          autoFocus
          required
        />

        <input
          className="form-group__text-input"
          type="text"
          placeholder="Amount (required)"
          value={this.state.amount}
          onChange={this.onAmountChange}
          required
        />

        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />

        <textarea
          className="form-group__textarea"
          name="notes"
          id=""
          cols="40"
          rows="3"
          placeholder="Additional Notes (optional)"
          value={this.state.note}
          onChange={this.onNoteChange}
        />

        <div>
          <button className="button form-group__submit-button">
            {this.props.actionText}
          </button>
        </div>
      </form>
    );
  }
}
