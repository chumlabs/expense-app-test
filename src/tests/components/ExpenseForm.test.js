import React from 'react';
import { shallow } from 'enzyme';
import ExpenseForm from '../../components/ExpenseForm';
import expenses, { now } from '../fixtures/expenseData';
// import { SingleDatePicker } from 'react-dates';

test('should render ExpenseForm (no data)', () => {
  const wrapper = shallow(<ExpenseForm />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseForm with expense data', () => {
  const wrapper = shallow(<ExpenseForm expense={expenses[0]} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid form data in ExpenseForm', () => {
  const wrapper = shallow(<ExpenseForm />);
  expect(wrapper).toMatchSnapshot(); // snapshot before submit
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {}
  });
  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot(); // snapshot after submit
});

test('should set decription on description input change in ExpenseForm', () => {
  const value = 'new description';
  const wrapper = shallow(<ExpenseForm />);
  wrapper
    .find('input')
    .at(0)
    .simulate('change', {
      target: { value }
    });
  expect(wrapper.state('description')).toBe(value);
});

test('should set note on note input change in ExpenseForm', () => {
  const value = 'new note value';
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find('textarea').simulate('change', {
    target: { value }
  });
  expect(wrapper.state('note')).toBe(value);
});

test('should set amount on amount input change using valid data', () => {
  const value = '7.04';
  const wrapper = shallow(<ExpenseForm />);
  wrapper
    .find('input')
    .at(1)
    .simulate('change', {
      target: { value }
    });
  expect(wrapper.state('amount')).toBe(value);
});

test('should not set amount on amount input change using invalid data', () => {
  const value = 'invalid data';
  const wrapper = shallow(<ExpenseForm />);
  wrapper
    .find('input')
    .at(1)
    .simulate('change', {
      target: { value }
    });
  expect(wrapper.state('amount')).toBe('');
});

test('should call onSubmit prop for valid form submission', () => {
  const onSubmitSpy = jest.fn();
  const wrapper = shallow(<ExpenseForm expense={expenses[0]} onSubmit={onSubmitSpy} />);
  const { description, amount, createdAt, note } = expenses[0];
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {}
  }); // simulate submit event
  expect(wrapper.state('error')).toBe(''); // check error state
  expect(onSubmitSpy).toHaveBeenLastCalledWith({ description, amount, createdAt, note });
});

test('should set new date on date change', () => {
  const wrapper = shallow(<ExpenseForm />);
  // wrapper.find('withStyles(SingleDatePicker)').prop('onDateChange')(now); // using display name
  wrapper.find('[onDateChange]').prop('onDateChange')(now); // using prop selector
  expect(wrapper.state('createdAt')).toEqual(now);
});

test('should set focused prop when onFocusChange is triggered', () => {
  const focused = true;
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find('[onDateChange]').prop('onFocusChange')({ focused }); // using prop selector
  // wrapper.find(SingleDatePicker).prop('onFocusChange')({ focused });  // using component (requires import)
  expect(wrapper.state('calendarFocused')).toBe(focused);
});
