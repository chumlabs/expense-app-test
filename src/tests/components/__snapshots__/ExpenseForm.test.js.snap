// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseForm (no data) 1`] = `
<form
  className="form-group"
  onSubmit={[Function]}
>
  <div
    className="form-group__error"
  />
  <input
    autoFocus={true}
    className="form-group__text-input"
    onChange={[Function]}
    placeholder="Description (required)"
    required={true}
    type="text"
    value=""
  />
  <input
    className="form-group__text-input"
    onChange={[Function]}
    placeholder="Amount (required)"
    required={true}
    type="text"
    value=""
  />
  <withStyles(SingleDatePicker)
    anchorDirection="left"
    block={false}
    calendarInfoPosition="bottom"
    customCloseIcon={null}
    customInputIcon={null}
    date={"1970-01-01T00:00:00.000Z"}
    daySize={39}
    disabled={false}
    displayFormat={[Function]}
    enableOutsideDays={false}
    firstDayOfWeek={null}
    focused={false}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    id="date"
    initialVisibleMonth={null}
    inputIconPosition="before"
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    monthFormat="MMMM YYYY"
    navNext={null}
    navPrev={null}
    noBorder={false}
    numberOfMonths={1}
    onClose={[Function]}
    onDateChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    openDirection="down"
    orientation="horizontal"
    phrases={
      Object {
        "calendarLabel": "Calendar",
        "chooseAvailableDate": [Function],
        "clearDate": "Clear Date",
        "closeDatePicker": "Close",
        "dateIsSelected": [Function],
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month.",
        "jumpToPrevMonth": "Move backward to switch to the previous month.",
        "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
        "moveFocusByOneMonth": "Switch months.",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
        "openThisPanel": "Open this panel.",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field.",
        "selectFocusedDate": "Select the date in focus.",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    placeholder="Date"
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonth={null}
    reopenPickerOnClearDate={false}
    required={false}
    screenReaderInputMessage=""
    showClearDate={false}
    showDefaultInputIcon={false}
    small={false}
    verticalHeight={null}
    verticalSpacing={22}
    weekDayFormat="dd"
    withFullScreenPortal={false}
    withPortal={false}
  />
  <textarea
    className="form-group__textarea"
    cols="40"
    id=""
    name="notes"
    onChange={[Function]}
    placeholder="Additional Notes (optional)"
    rows="3"
    value=""
  />
  <div>
    <button
      className="button form-group__submit-button"
    />
  </div>
</form>
`;

exports[`should render ExpenseForm with expense data 1`] = `
<form
  className="form-group"
  onSubmit={[Function]}
>
  <div
    className="form-group__error"
  />
  <input
    autoFocus={true}
    className="form-group__text-input"
    onChange={[Function]}
    placeholder="Description (required)"
    required={true}
    type="text"
    value="item 1 - abc"
  />
  <input
    className="form-group__text-input"
    onChange={[Function]}
    placeholder="Amount (required)"
    required={true}
    type="text"
    value="7.00"
  />
  <withStyles(SingleDatePicker)
    anchorDirection="left"
    block={false}
    calendarInfoPosition="bottom"
    customCloseIcon={null}
    customInputIcon={null}
    date={"2017-04-01T16:00:00.000Z"}
    daySize={39}
    disabled={false}
    displayFormat={[Function]}
    enableOutsideDays={false}
    firstDayOfWeek={null}
    focused={false}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    id="date"
    initialVisibleMonth={null}
    inputIconPosition="before"
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    monthFormat="MMMM YYYY"
    navNext={null}
    navPrev={null}
    noBorder={false}
    numberOfMonths={1}
    onClose={[Function]}
    onDateChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    openDirection="down"
    orientation="horizontal"
    phrases={
      Object {
        "calendarLabel": "Calendar",
        "chooseAvailableDate": [Function],
        "clearDate": "Clear Date",
        "closeDatePicker": "Close",
        "dateIsSelected": [Function],
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month.",
        "jumpToPrevMonth": "Move backward to switch to the previous month.",
        "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
        "moveFocusByOneMonth": "Switch months.",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
        "openThisPanel": "Open this panel.",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field.",
        "selectFocusedDate": "Select the date in focus.",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    placeholder="Date"
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonth={null}
    reopenPickerOnClearDate={false}
    required={false}
    screenReaderInputMessage=""
    showClearDate={false}
    showDefaultInputIcon={false}
    small={false}
    verticalHeight={null}
    verticalSpacing={22}
    weekDayFormat="dd"
    withFullScreenPortal={false}
    withPortal={false}
  />
  <textarea
    className="form-group__textarea"
    cols="40"
    id=""
    name="notes"
    onChange={[Function]}
    placeholder="Additional Notes (optional)"
    rows="3"
    value="lowest amount, middle date"
  />
  <div>
    <button
      className="button form-group__submit-button"
    />
  </div>
</form>
`;

exports[`should render error for invalid form data in ExpenseForm 1`] = `
<form
  className="form-group"
  onSubmit={[Function]}
>
  <div
    className="form-group__error"
  />
  <input
    autoFocus={true}
    className="form-group__text-input"
    onChange={[Function]}
    placeholder="Description (required)"
    required={true}
    type="text"
    value=""
  />
  <input
    className="form-group__text-input"
    onChange={[Function]}
    placeholder="Amount (required)"
    required={true}
    type="text"
    value=""
  />
  <withStyles(SingleDatePicker)
    anchorDirection="left"
    block={false}
    calendarInfoPosition="bottom"
    customCloseIcon={null}
    customInputIcon={null}
    date={"1970-01-01T00:00:00.000Z"}
    daySize={39}
    disabled={false}
    displayFormat={[Function]}
    enableOutsideDays={false}
    firstDayOfWeek={null}
    focused={false}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    id="date"
    initialVisibleMonth={null}
    inputIconPosition="before"
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    monthFormat="MMMM YYYY"
    navNext={null}
    navPrev={null}
    noBorder={false}
    numberOfMonths={1}
    onClose={[Function]}
    onDateChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    openDirection="down"
    orientation="horizontal"
    phrases={
      Object {
        "calendarLabel": "Calendar",
        "chooseAvailableDate": [Function],
        "clearDate": "Clear Date",
        "closeDatePicker": "Close",
        "dateIsSelected": [Function],
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month.",
        "jumpToPrevMonth": "Move backward to switch to the previous month.",
        "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
        "moveFocusByOneMonth": "Switch months.",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
        "openThisPanel": "Open this panel.",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field.",
        "selectFocusedDate": "Select the date in focus.",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    placeholder="Date"
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonth={null}
    reopenPickerOnClearDate={false}
    required={false}
    screenReaderInputMessage=""
    showClearDate={false}
    showDefaultInputIcon={false}
    small={false}
    verticalHeight={null}
    verticalSpacing={22}
    weekDayFormat="dd"
    withFullScreenPortal={false}
    withPortal={false}
  />
  <textarea
    className="form-group__textarea"
    cols="40"
    id=""
    name="notes"
    onChange={[Function]}
    placeholder="Additional Notes (optional)"
    rows="3"
    value=""
  />
  <div>
    <button
      className="button form-group__submit-button"
    />
  </div>
</form>
`;

exports[`should render error for invalid form data in ExpenseForm 2`] = `
<form
  className="form-group"
  onSubmit={[Function]}
>
  <div
    className="form-group__error"
  >
    <p
      className="error-text"
    >
      Please enter a description
    </p>
  </div>
  <input
    autoFocus={true}
    className="form-group__text-input"
    onChange={[Function]}
    placeholder="Description (required)"
    required={true}
    type="text"
    value=""
  />
  <input
    className="form-group__text-input"
    onChange={[Function]}
    placeholder="Amount (required)"
    required={true}
    type="text"
    value=""
  />
  <withStyles(SingleDatePicker)
    anchorDirection="left"
    block={false}
    calendarInfoPosition="bottom"
    customCloseIcon={null}
    customInputIcon={null}
    date={"1970-01-01T00:00:00.000Z"}
    daySize={39}
    disabled={false}
    displayFormat={[Function]}
    enableOutsideDays={false}
    firstDayOfWeek={null}
    focused={false}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    id="date"
    initialVisibleMonth={null}
    inputIconPosition="before"
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    monthFormat="MMMM YYYY"
    navNext={null}
    navPrev={null}
    noBorder={false}
    numberOfMonths={1}
    onClose={[Function]}
    onDateChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    openDirection="down"
    orientation="horizontal"
    phrases={
      Object {
        "calendarLabel": "Calendar",
        "chooseAvailableDate": [Function],
        "clearDate": "Clear Date",
        "closeDatePicker": "Close",
        "dateIsSelected": [Function],
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month.",
        "jumpToPrevMonth": "Move backward to switch to the previous month.",
        "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
        "moveFocusByOneMonth": "Switch months.",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
        "openThisPanel": "Open this panel.",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field.",
        "selectFocusedDate": "Select the date in focus.",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    placeholder="Date"
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonth={null}
    reopenPickerOnClearDate={false}
    required={false}
    screenReaderInputMessage=""
    showClearDate={false}
    showDefaultInputIcon={false}
    small={false}
    verticalHeight={null}
    verticalSpacing={22}
    weekDayFormat="dd"
    withFullScreenPortal={false}
    withPortal={false}
  />
  <textarea
    className="form-group__textarea"
    cols="40"
    id=""
    name="notes"
    onChange={[Function]}
    placeholder="Additional Notes (optional)"
    rows="3"
    value=""
  />
  <div>
    <button
      className="button form-group__submit-button"
    />
  </div>
</form>
`;
